kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/ConfigSourcePrecedenceTest.java

```diff
@@ -0,0 +1,220 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client;
+
+import io.fabric8.kubernetes.client.utils.Utils;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+class ConfigSourcePrecedenceTest {
+
+  @Nested
+  @DisplayName("kubeconfig")
+  class KubeConfigSource {
+    @BeforeEach
+    void setUp() {
+      System.setProperty("kubeconfig", Utils.filePath(ConfigSourcePrecedenceTest.class.getResource("/test-kubeconfig")));
+    }
+
+    @Test
+    @DisplayName("no other source provided, use kubeconfig attributes in Config")
+    void whenNoOtherSourceProvided_thenUseKubeConfig() {
+      // Given
+      Config config = new ConfigBuilder().build();
+
+      // When + Then
+      assertThat(config)
+          .hasFieldOrPropertyWithValue("masterUrl", "https://172.28.128.4:8443/")
+          .hasFieldOrPropertyWithValue("namespace", "testns")
+          .hasFieldOrPropertyWithValue("autoOAuthToken", "token");
+    }
+
+    @Nested
+    @DisplayName("user configuration overrides KubeConfig attributes")
+    class UserConfigurationOverKubeConfig extends UserConfigurationViaConfigBuilder {
+    }
+
+    @Test
+    @DisplayName("System Properties configured, then give precedence to System Properties")
+    void whenSystemPropertiesUsedForSomeFields_thenSystemPropertiesGivenPrecedence() {
+      try {
+        // Given
+        System.setProperty("kubernetes.master", "https://user-configuration-override:8443");
+        System.setProperty("kubernetes.namespace", "namespace-overridden");
+        System.setProperty("kubernetes.auth.token", "token-overridden");
+
+        // When
+        Config config = new ConfigBuilder().build();
+
+        // Then
+        assertThat(config)
+            .hasFieldOrPropertyWithValue("masterUrl", "https://user-configuration-override:8443/")
+            .hasFieldOrPropertyWithValue("namespace", "namespace-overridden")
+            .hasFieldOrPropertyWithValue("autoOAuthToken", "token-overridden");
+      } finally {
+        System.clearProperty("kubernetes.master");
+        System.clearProperty("kubernetes.namespace");
+        System.clearProperty("kubernetes.auth.token");
+      }
+    }
+
+    @Test
+    @DisplayName("Service Account files provided, then do NOT use it")
+    void whenServiceAccountPropertyConfigured_thenDoNotUseIt() {
+      try {
+        // Given
+        System.setProperty("KUBERNETES_SERVICE_HOST", "10.96.0.1");
+        System.setProperty("KUBERNETES_SERVICE_PORT", "443");
```

Comment:
Why do this look like environment variables?

Reply:
Yes, Config class uses `Utils.getSystemPropertyOrEnvVar` to resolve kubernetes apiserver host and port

https://github.com/fabric8io/kubernetes-client/blob/520894adcb24913998d0792b3921b71e1b76c86a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java#L531-L532

This method can handle converting property name `kubernetes.service.host`  to environment variable `KUBERNETES_SERVICE_HOST`; but I think we're passing it incorrectly. At the moment, it only seems to work as both property and env names as `KUBERNETES_SERVICE_HOST` 

https://github.com/fabric8io/kubernetes-client/blob/520894adcb24913998d0792b3921b71e1b76c86a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java#L121-L122

Reply:
So maybe report this as a bug?

Also, given your context, not sure if this is the valid test name then?

Reply:
Umm, I'm not sure if it's a bug as KubernetesClient doesn't document anywhere usage of `kubernetes.service.host` and `kubernetes.service.port` properties in https://github.com/fabric8io/kubernetes-client#configuring-the-client

`KUBERNETES_SERVICE_HOST/PORT`  environment variables are something that's exposed via Kubernetes. KubernetesClient is just trying to detect it in auto configuration phase.

Reply:
I have removed adding `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` to System Properties as they're not System Properties.